File: go/array/108_sorted_array_to_BST.go
// Time Complexity: O(n)
// Space Complexity: O(n)
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sortedArrayToBST(nums []int) *TreeNode {
	return numsToBST(nums, 0, len(nums)-1)
}

func numsToBST(nums []int, left, right int) *TreeNode {
	if left > right {
		return nil
	}

	mid := left + (right-left)/2

	node := &TreeNode{}

	node.Val = nums[mid]
	node.Left = numsToBST(nums, left, mid-1)
	node.Right = numsToBST(nums, mid+1, right)

	return node
}

File: python/array/108_sorted_array_to_BST.py
# Time Complexity: O(n)
# Space Complexity: O(n)

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

from typing import List, Optional

class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def numsToBST(left: int, right: int) -> Optional[TreeNode]:
            if left > right:
                return None
            
            mid = left + (right - left) // 2
            
            node = TreeNode(nums[mid])
            node.left = numsToBST(left, mid - 1)
            node.right = numsToBST(mid + 1, right)
            
            return node
        
        return numsToBST(0, len(nums) - 1)